go/v1beta1/server/server.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2019 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package to include all generic instantiation of the Grafeas server, designed to be called from a very small main() implementation.
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/cockroachdb/cmux"
+	"github.com/grafeas/grafeas/go/config"
+	"github.com/grafeas/grafeas/go/v1beta1/api"
+	"github.com/grafeas/grafeas/go/v1beta1/project"
+	"github.com/grafeas/grafeas/go/v1beta1/storage"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	prpb "github.com/grafeas/grafeas/proto/v1beta1/project_go_proto"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/rs/cors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+var (
+	configFile = flag.String("config", "", "Path to a config file")
+)
+
+// Start the Grafeas server, instantiating the storage of the type specified in the config.
+func StartGrafeas() error {
+	flag.Parse()
+	cfg, err := config.LoadConfig(*configFile)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Failed to load cfg file: %s", err))
+	}
+	var db grafeas.Storage
+	var proj project.Storage
+
+	s, err := storage.CreateStorageOfType(cfg.StorageType, cfg.StorageConfig)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Failed to create storage: %s", err))
+	}
+	db = s
+	proj = s
+	run(cfg.API, &db, &proj)
+	return nil
+}
+
+// run initializes grpc and grpc gateway api services on the same address
+func run(config *config.ServerConfig, db *grafeas.Storage, proj *project.Storage) {
+	network, address := "tcp", config.Address
+	if strings.HasPrefix(config.Address, "unix://") {
+		network = "unix"
+		address = strings.TrimPrefix(config.Address, "unix://")
+		// Remove existing socket if found
+		os.Remove(address)
+	}
+	l, err := net.Listen(network, address)
+	if err != nil {
+		log.Fatalln("could not listen to address", config.Address)
```

Comment:
I mentioned this before, but typically we only log.Fatal in main, it's more idiomatic for all other functions to return errors instead, generally so they can be tested.

Reply:
Amended.  Note that none of these functions are mine - they're your existing codebase - but I have amended to return errors.